File: frontend/src/types/report.ts
// v2 types for the Supervisor-led report system.
// These coexist with legacy task-centric types during the migration.

export type ReportType =
  | 'ops_review'
  | 'internal_research'
  | 'risk_assessment'
  | 'regulatory_filing'
  | 'training_material'
  // User-defined types take the form `custom:<id>`.
  | (string & { readonly __brand?: 'custom' });

export type ReportStatus =
  | 'draft'
  | 'intake'
  | 'scoping'
  | 'producing'
  | 'reviewing'
  | 'delivered'
  | 'failed'
  | 'cancelled';

export type ReportPhase = ReportStatus | string;

export type MessageRole =
  | 'supervisor_say'
  | 'supervisor_ask'
  | 'team_change'
  | 'employee_note'
  | 'user_reply'
  | 'user_interject'
  | 'phase_transition';

export interface ReportMessage {
  id: number;
  report_id: number;
  role: MessageRole;
  author_id?: string | null;
  author_name?: string | null;
  content: string;
  meta?: Record<string, any> | null;
  created_at: string;
}

export interface Clarification {
  id: number;
  report_id: number;
  question: string;
  default_answer?: string | null;
  answer?: string | null;
  status: 'pending' | 'answered' | 'defaulted' | 'skipped';
  priority?: 'high' | 'medium' | 'low';
  created_at: string;
  answered_at?: string | null;
}

export interface TimelineEvent {
  id: number;
  report_id: number;
  event_type: string;
  label: string;
  payload?: Record<string, any> | null;
  created_at: string;
}

export interface SectionOutlineItem {
  id: string;
  title: string;
  kind: string;
}

export interface OutputSection {
  text: string;
  employee_id?: string;
  employee_name?: string;
  note?: string;
  error?: string | null;
}

export interface Report {
  id: number;
  user_id: number;
  title: string;
  brief: string;
  report_type: ReportType;
  depth: string;
  status: ReportStatus;
  phase: ReportPhase;
  progress: number;
  scoping_plan?: any;
  team_roster?: string[] | null;
  section_outline?: SectionOutlineItem[] | null;
  output_index?: Record<string, OutputSection> | null;
  final_file_path?: string | null;
  final_file_name?: string | null;
  error_message?: string | null;
  trace_log?: Record<string, any> | null;
  data_context?: Record<string, any> | null;
  created_at: string;
  updated_at: string;
  started_at?: string | null;
  completed_at?: string | null;
}

export interface ReportDetail extends Report {
  messages: ReportMessage[];
  clarifications: Clarification[];
  timeline: TimelineEvent[];
}

export interface ReportTypeInfo {
  id: ReportType;
  label: string;
  label_en: string;
  description: string;
  typical_inputs: string[];
  typical_output: string;
  default_team: string[];
  section_skeleton: SectionOutlineItem[];
  is_custom?: boolean;
  visibility?: 'private' | 'public';
  owner_is_me?: boolean;
}

// ----------------------------- Workforce -----------------------------

export interface WorkforceMember {
  id: string;
  name: string;
  first_name_en: string;
  role_title_en: string;
  tagline_en: string;
  portrait_seed: string;
  category: string;
  description: string;
  skills: string[];
  tools: string[];
  applicable_report_types: string[];
  inputs: string[];
  outputs: string[];
  default_model?: string | null;
  resolved_model_id?: string | null;
  resolved_model_name?: string | null;
  enabled: boolean;
  is_supervisor?: boolean;
}

export interface WorkforceResponse {
  supervisor: WorkforceMember;
  employees: WorkforceMember[];
}

// ----------------------------- API Keys ------------------------------

export interface ApiKeyRecord {
  id: number;
  user_id: number;
  name: string;
  prefix: string;
  masked_key: string;
  status: 'active' | 'suspended' | 'revoked';
  daily_request_limit?: number | null;
  monthly_token_limit?: number | null;
  total_requests: number;
  last_used_at?: string | null;
  created_at: string;
  expires_at?: string | null;
  raw_key?: string; // Only present on creation
}

File: backend/app/models/report.py
from datetime import datetime
from sqlalchemy import String, Integer, Float, JSON, Text, DateTime, ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column
from app.database import Base


class Report(Base):
    __tablename__ = "reports"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), index=True)
    api_key_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("api_keys.id"), index=True, nullable=True)
    title: Mapped[str] = mapped_column(String(500))
    brief: Mapped[str] = mapped_column(Text)
    report_type: Mapped[str] = mapped_column(String(17), index=True)
    depth: Mapped[str] = mapped_column(String(20), default="standard")
    status: Mapped[str] = mapped_column(String(9), default="pending", index=True)  # pending, running, completed, failed
    progress: Mapped[float] = mapped_column(Float, default=0.0)
    phase: Mapped[str] = mapped_column(String(40), default="初始化")
    scoping_plan: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    team_roster: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    section_outline: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    output_index: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    final_file_path: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    final_file_name: Mapped[str | None] = mapped_column(String(500), nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    output_format: Mapped[str] = mapped_column(String(20), default="word")
    template_file_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    trace_log: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    data_context: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=func.now(), index=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=func.now(), onupdate=func.now())
    started_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
